src/core/tsi/ssl_transport_security_utils.cc

```diff
@@ -375,4 +380,79 @@ absl::StatusOr<std::string> AkidFromCrl(X509_CRL* crl) {
   return ret;
 }
 
+bool IsPemCertificateChainNonEmptyAndValid(absl::string_view cert_chain_pem) {
+  if (cert_chain_pem.empty()) {
+    return false;
+  }
+  BIO* in = BIO_new_mem_buf(cert_chain_pem.data(), cert_chain_pem.size());
+  if (in == nullptr) {
+    return false;
+  }
+  uint8_t* cert_data = nullptr;
+  long cert_length = 0;
+  size_t number_of_valid_certs = 0;
+  while (PEM_bytes_read_bio(&cert_data, &cert_length, /*pnm=*/nullptr,
+                            PEM_STRING_X509, in, /*cb=*/nullptr,
+                            /*u=*/nullptr)) {
+    // Try to form the X.509 object from the DER-encoding.
+    const uint8_t* der = cert_data;
+    X509* x509 = d2i_X509(/*out=*/nullptr, &der, cert_length);
+    if (x509 == nullptr) {
+      BIO_free(in);
+      return false;
+    }
+    X509_free(x509);
+    OPENSSL_free(cert_data);
+    cert_data = nullptr;
+    cert_length = 0;
+    number_of_valid_certs++;
+  }
+  // We always have errors at this point because in the above loop we read until
+  // we reach the end of |cert_chain_pem|, which generates a "no start line"
+  // error. Therefore, this error is OK if we have successfully parsed some
+  // certificate data previously.
+  const int last_error = ERR_peek_last_error();
+  if (ERR_GET_LIB(last_error) != ERR_LIB_PEM ||
+      ERR_GET_REASON(last_error) != PEM_R_NO_START_LINE) {
+    BIO_free(in);
```

Comment:
This works, but given that you're C++, do you all have smart pointers for OpenSSL types? If you're depending on BoringSSL, you could use `bssl::UniquePtr<BIO>`, but I'm guessing that won't work with OpenSSL.

Reply:
Unfortunately, I don't think we can use `bssl::UniquePtr` as we need to be compatible with OpenSSL. (As an aside, adding our own smart pointer types is a good idea so we can avoid all the manual cleanup, but I'll leave that out-of-scope for this PR if that's OK.)